### PR TITLE
State

### LINKED ISSUE #8463: Mapcap should save also states of the widgets

### LINKED ISSUE #8464: State of the widgets is not 'persisted' on the builder.
STR

Create a map
Add a widget histogram
Filter
Go to the dashboard
Go back to the map.

The widget should be filtering but it's not.

### PR DESCRIPTION
**Pull request by gfiorav**
Fixes #8463
Fixes #8464

# Synopsis

This makes state (which had already been seen at deep insights level) a standard trade coin for visualisations in the back end. There is now a State model, whose spirit is to capture the current state of a given visualisation (its widget's filters and viewport ATM). 

For example, if you apply a filter in a widget and refresh the builder view, the same applied state will be reloaded into the the deep insights dashboard (so everything is exactly how you left it). In addition, when the vis is shared, any state will be applied in the embed as you left it or map capped it.

State is exportable as well.

## State model

```ruby
Uuid     :id, primary_key: true, default: 'uuid_generate_v4()'.lit
DateTime :created_at, default: Sequel::CURRENT_TIMESTAMP
DateTime :updated_at, default: Sequel::CURRENT_TIMESTAMP
String   :json, null: false, type: 'json'

foreign_key :user_id, :users, type: :uuid, null: false, on_delete: :cascade
foreign_key :visualization_id, :visualizations, type: :uuid, null: false, on_delete: :cascade

add_index [:user_id, :visualization_id]
```

This model is associated to a vis and a user (see snapshots feature design in consideration). In this implementation, there is one, global, state for evert vis that can only be updated (`.../vis/:id/state`). Updates are triggered by viewport or widget changes.

## Considerations

It should be considered that this same model will be reused for upcoming snapshots feature, by which different users will be able to save different embed states and interact with them (like 10 per user/vis). Although it will use the same model, it will be an entirely different endpoint (a normal CRUD).

Please review the front end @xavijam 
Please review the back end @javitonino and @juanignaciosl (in that order, I expect a lot of design comments.

### CODE REVIEW

File: app/services/carto/visualizations_export_service_2.rb [hunk 1]
```diff
@@ -9,7 +9,7 @@
 # 2.0.2: export username
```
Comment (javitonino) [inline app/services/carto/visualizations_export_service_2.rb]: A good place to add a comment describing what 2.0.3 does.

File: db/migrate/20160808114841_create_states.rb [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+      DateTime :created_at, default: Sequel::CURRENT_TIMESTAMP
+      DateTime :updated_at, default: Sequel::CURRENT_TIMESTAMP
+      String   :json, null: false, type: 'json'
+
+      foreign_key :user_id, :users, type: :uuid, null: false, on_delete: :cascade
+      foreign_key :visualization_id, :visualizations, type: :uuid, null: false, on_delete: :cascade
+    end
+
+    alter_table :states do
+      add_index [:user_id, :visualization_id]
```
Comment (javitonino) [inline db/migrate/20160808114841_create_states.rb]: What is more likely to be requested in the future:
 - The states of an user
 - The states of a visualization

This index is only effective for the case both columns are specified or only `user_id` is given. Keep in mind that deleting (cascade) needs to run this query. So, if I got my DB's right, deleting a visualization will do a full scan of the state table to delete related entities.

Maybe it's better to have two indices (one for each column). Can you do some testing (explains?)

File: db/migrate/20160808114842_add_state_id_to_visualizations.rb [hunk 3]
```diff
@@ -0,0 +1,14 @@
+Sequel.migration do
+  up do
+    alter_table :visualizations do
+      add_foreign_key :state_id, :states, type: 'uuid', null: true
```
Comment (javitonino) [inline db/migrate/20160808114842_add_state_id_to_visualizations.rb]: This one should be fine (I run a similar one this morning, no problems). But remember to split each migration to its own PR for safety :fire_engine:

File: app/models/carto/visualization.rb [hunk 4]
```diff
@@ -149,7 +154,7 @@ def is_publically_accesible?
     is_public? || is_link_privacy?
   end
 
-  def is_writable_by_user(user)
+  def writable_by?(user)
```
Comment (javitonino) [inline app/models/carto/visualization.rb]: :sparkles:

File: app/models/carto/visualization.rb [hunk 5]
```diff
@@ -386,6 +391,11 @@ def add_source_analyses
     end
   end
 
+  def state_with_creation
+    state_without_creation || create_state(user_id: user_id)
```
Comment (javitonino) [inline app/models/carto/visualization.rb]: `create_state` persists, `build_state` does not. Do you need `create`? For user notifications I didn't, as I only saved the actual notification if it was modified, no need to save a blank one.

File: app/controllers/carto/api/state_presenter.rb [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+        @state = state
+      end
+
+      def to_hash
+        {
+          id: @state.id,
+          created_at: @state.created_at,
+          updated_at: @state.updated_at,
+          json: @state.json,
+          user: Carto::Api::UserPresenter.new(@state.user).to_public_poro
```
Comment (javitonino) [inline app/controllers/carto/api/state_presenter.rb]: Maybe overkill, but `to_public_poro` is not too bad, so :+1:

File: app/controllers/carto/api/states_controller.rb [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+
+      def load_visualization
+        visualization_id = params[:visualization_id]
+
+        @visualization = Carto::Visualization.where(id: visualization_id).first if visualization_id
+        raise Carto::LoadError.new("Visualization not found: #{visualization_id}") unless @visualization
+      end
+
+      def check_writer
+        @visualization.writable_by?(current_viewer)
```
Comment (javitonino) [inline app/controllers/carto/api/states_controller.rb]: Does this work? Shouldn't it raise `Carto::UnauthorizedError`?

File: app/controllers/carto/builder/public/embeds_controller.rb [hunk 8]
```diff
@@ -35,6 +37,13 @@ def load_visualization
...
         end
 
+        def load_visualization_for_vizjson_and_state
+          @visualization_for_vizjson_and_state = if @visualization.mapcapped?
+                                                   @visualization.latest_mapcap.regenerate_visualization
+                                                 else
+                                                   @visualization
+                                                 end
+
         def load_auth_tokens
```
Comment (javitonino) [inline app/controllers/carto/builder/public/embeds_controller.rb]: dat missing `end` breaking the tests

File: app/models/carto/map.rb [hunk 9]
```diff
@@ -96,7 +96,7 @@ def view_bounds_data
   end
 
   def writable_by_user?(user)
-    !visualizations.select { |v| v.is_writable_by_user(user) }.empty?
+    !visualizations.select { |v| v.writable_by?(user) }.empty?
```
Comment (javitonino) [inline app/models/carto/map.rb]: Maybe `!xxx.empty?` => `xxx.any?`

File: app/models/carto/state.rb [hunk 10]
```diff
@@ -0,0 +1,24 @@
+# encoding: UTF-8
+
+require 'json'
+require_relative './carto_json_serializer'
+
+class Carto::State < ActiveRecord::Base
+  belongs_to :visualization, class_name: Carto::Visualization
+  belongs_to :user, class_name: Carto::User
+
+  default_scope order('created_at DESC')
```
Comment (javitonino) [inline app/models/carto/state.rb]: After my latest experience with `default_scope` I don't trust it, but it should be good for something like this where it always need to be applied.

File: db/migrate/20160808114841_create_states.rb [hunk 11]
```diff
@@ -0,0 +1,21 @@
+Sequel.migration do
+  up do
+    create_table :states do
+      Uuid     :id, primary_key: true, default: 'uuid_generate_v4()'.lit
+      DateTime :created_at, default: Sequel::CURRENT_TIMESTAMP
+      DateTime :updated_at, default: Sequel::CURRENT_TIMESTAMP
+      String   :json, null: false, type: 'json'
```
Comment (javitonino) [inline db/migrate/20160808114841_create_states.rb]: As this is mandatory and you initialize it if needed in Rails, consider adding a default to the DB too.

File: app/models/carto/visualization.rb [hunk 12]
```diff
@@ -34,6 +34,9 @@ class Carto::Visualization < ActiveRecord::Base
 
   belongs_to :permission
 
+  has_one :state, class_name: Carto::State
+  has_many :viewers, class_name: Carto::User
```
Comment (javitonino) [inline app/models/carto/visualization.rb]: This is not used right now, so I'd remove it. We can add it when needed, maybe thinking of a different name to avoid confusion with viewer/builder users.

File: app/controllers/carto/builder/public/embeds_controller.rb (hunk 8 above)
Comment (gfiorav) [inline app/controllers/carto/builder/public/embeds_controller.rb]: Damn, resolved the conflict wrong 🍨

File: app/services/carto/visualizations_export_persistence_service.rb [hunk 13]
```diff
@@ -59,6 +59,11 @@ def save_import(user, visualization, renamed_tables: {})
         map.data_layers.each(&:register_table_dependencies)
       end
 
+      state = visualization.state
```
Comment (javitonino) [inline app/services/carto/visualizations_export_persistence_service.rb]: Can `state` be nil here?

File: config/routes.rb [hunk 14]
```diff
@@ -495,6 +495,7 @@
       scope '/viz/:visualization_id', constraints: { id: /[^\/]+/ } do
         resources :analyses, only: [:show, :create, :update, :destroy], constraints: { id: /[^\/]+/ }
         resources :mapcaps, only: [:index, :show, :create, :destroy], constraints: { id: /[^\/]+/ }
+        resource :state, only: [:update]
```
Comment (javitonino) [inline config/routes.rb]: Will you be using `resources` for the controllers for the user-viz state, right? Both will map to the same controller:
```
Because you might want to use the same controller for a singular route (/account) and a plural route (/accounts/45), singular resources map to plural controllers. So that, for example, resource :photo and resources :photos creates both singular and plural routes that map to the same controller (PhotosController).
```

Just to keep it in mind for the future, not very relevant for this PR.

**Comment by javitonino** [comment]
Left a bunch of comments! Will give you a chance to take a look at them before continuing. Nothing too serious though, this is looking nice :palm_tree:

File: db/migrate/20160808114841_create_states.rb (hunk 2 above)
Comment (gfiorav) [inline db/migrate/20160808114841_create_states.rb]: But specifying both is exactly the most used case right?

I'm scared with cascading now.

File: app/models/carto/visualization.rb (hunk 5 above)
Comment (gfiorav) [inline app/models/carto/visualization.rb]: Yup!

I like it this way 😁

File: app/controllers/carto/api/state_presenter.rb (hunk 6 above)
Comment (gfiorav) [inline app/controllers/carto/api/state_presenter.rb]: Meh, I think we're safe. it's already public safe.

File: app/controllers/carto/api/states_controller.rb (hunk 7 above)
Comment (gfiorav) [inline app/controllers/carto/api/states_controller.rb]: Why would it? Not from my testing :/

File: app/models/carto/map.rb (hunk 9 above)
Comment (gfiorav) [inline app/models/carto/map.rb]: (I just changed the method name)

Yeah, it's better, I'll change it!

File: db/migrate/20160808114841_create_states.rb (hunk 2 above)
Comment (javitonino) [inline db/migrate/20160808114841_create_states.rb]: Yes, it is for explicit queries. I don't know if there will be a need in the future for queries by user or viz. But we also have the implicit "queries" run by postgresql when cascading, which must be able to identify which rows to delete. 

So deleting a visualization will trigger something like "DELETE FROM state WHERE visualization_id = <deleted_id>". This query has no index that it can use -> bad times.

Removing the cascade doesn't totally solve the problem, as the FK is still there, so deleting a visualization must delete all of its states one way or another, or it will fail.

File: app/models/carto/state.rb (hunk 10 above)
Comment (gfiorav) [inline app/models/carto/state.rb]: Fingers crossed! 

I used a class method for Mapcaps. Do you think that's better?

File: db/migrate/20160808114841_create_states.rb (hunk 11 above)
Comment (gfiorav) [inline db/migrate/20160808114841_create_states.rb]: Ah, nice observation.

Shouldn't it make things blow up at db level? Don't feel confutable with it :/

File: app/models/carto/visualization.rb (hunk 12 above)
Comment (gfiorav) [inline app/models/carto/visualization.rb]: This is from an example.... AND it's duplicated also...

Changing 😓

File: app/services/carto/visualizations_export_persistence_service.rb (hunk 13 above)
Comment (gfiorav) [inline app/services/carto/visualizations_export_persistence_service.rb]: It's created if needed on instantiation, so I'd say no.

File: app/controllers/carto/api/states_controller.rb (hunk 7 above)
Comment (javitonino) [inline app/controllers/carto/api/states_controller.rb]: How does it map this response to 403 then? I guess some Rails magic default. I guess I expected to see `Carto::UnauthorizedError` from looking at the `rescue_from`, but can't argue if it works.

File: config/routes.rb (hunk 14 above)
Comment (gfiorav) [inline config/routes.rb]: Nah, the other endpoint will be named `/snapshots` but that's a very good point!

File: app/models/carto/map.rb (hunk 9 above)
Comment (javitonino) [inline app/models/carto/map.rb]: Troll Hound doesn't care about your puny notions of what has been changed. Obey! :stuck_out_tongue:

File: app/models/carto/state.rb (hunk 10 above)
Comment (javitonino) [inline app/models/carto/state.rb]: Kind of depends on how queries are built. I'd keep the `default_scope` here if no problems arise. By the way, this is an unindexed sort. Should not be a problem since states per user/viz will be limited though, but something to keep in mind if we decide to allow unlimited ones.

File: db/migrate/20160808114841_create_states.rb (hunk 2 above)
Comment (gfiorav) [inline db/migrate/20160808114841_create_states.rb]: Ok, I'm full scared now. 2 separate indexes then. Won't it be too costly?

File: app/services/carto/visualizations_export_persistence_service.rb (hunk 13 above)
Comment (javitonino) [inline app/services/carto/visualizations_export_persistence_service.rb]: Oh, true! Will do some unnecessary assignments in that case (user and visualization will be set if taking it from the db), but no biggie. This way, all newly imported visualization have some state set, which is cool.

File: app/controllers/carto/api/states_controller.rb (hunk 7 above)
Comment (gfiorav) [inline app/controllers/carto/api/states_controller.rb]: Ah, well that is good question. I have no clue how it knows to return 403.

File: db/migrate/20160808114841_create_states.rb (hunk 11 above)
Comment (javitonino) [inline db/migrate/20160808114841_create_states.rb]: Blow up how?

I like moving this kind of defaults to the lowest level possible, but it all works the same, so up to you.

File: db/migrate/20160808114841_create_states.rb (hunk 2 above)
Comment (javitonino) [inline db/migrate/20160808114841_create_states.rb]: I think it's fine, and index merging should do a good work for the default query, but let's bring in @juanignaciosl on this one.

File: app/models/carto/map.rb (hunk 9 above)
Comment (gfiorav) [inline app/models/carto/map.rb]: ✋ 😨

File: db/migrate/20160808114841_create_states.rb (hunk 11 above)
Comment (gfiorav) [inline db/migrate/20160808114841_create_states.rb]: ignore that, I was thinking of the `state_id` migration.

File: db/migrate/20160808114841_create_states.rb (hunk 2 above)
Comment (juanignaciosl) [inline db/migrate/20160808114841_create_states.rb]: [Let's summon Markus Winand on this](http://use-the-index-luke.com/sql/where-clause/searching-for-ranges/index-merge-performance):

> It is one of the most common question about indexing: is it better to create one index for each column or a single index for all columns of a where clause? The answer is very simple in most cases: one index with multiple columns is better—that is, a concatenated or compound index. [“Concatenated Indexes” explains them in detail](http://use-the-index-luke.com/sql/where-clause/the-equals-operator/concatenated-keys).

I won't C&P more text from those two indexes. They worth the read. Invest a few minutes on them and continue then.





Ready? :-)

I can't find queries of state by user in the code, could you tell me when would it be used?

Comment (javitonino) [inline db/migrate/20160808114841_create_states.rb]: Delete cascade of an user. It is not an explicit query but postgres must do something very similar to be able to say which states to delete when deleting the user.

Comment (juanignaciosl) [inline db/migrate/20160808114841_create_states.rb]: User deletion is not something that we should be worried about optimizing, specially if it's actually handled by PG. If that's the only reason, I'd only index by `visualization_id`. If you still have doubts, go for `[visualization_id, user_id]` for sure, but I wouldn't do it.

File: db/migrate/20160808114841_create_states.rb [hunk 15]
```diff
@@ -0,0 +1,21 @@
+Sequel.migration do
+  up do
+    create_table :states do
+      Uuid     :id, primary_key: true, default: 'uuid_generate_v4()'.lit
```
Comment (juanignaciosl) [inline db/migrate/20160808114841_create_states.rb]: First things first: if this is meant to support Mapcaps, shouldn't it have a `mapcap_id` (nullable) column? I'm not a big fan of adding things in advance anyway so, if you're splitting the feature first for visualizations, then for mapcaps, it's fine.

File: app/controllers/carto/api/mapcaps_controller.rb [hunk 16]
```diff
@@ -58,7 +58,7 @@ def load_visualization
       end
 
       def owners_only
-        raise Carto::UnauthorizedError.new unless @visualization.is_writable_by_user(current_user)
+        raise Carto::UnauthorizedError.new unless @visualization.writable_by?(current_user)
```
Comment (juanignaciosl) [inline app/controllers/carto/api/mapcaps_controller.rb]: Consider `writeable_by_user?` instead. Keeping `_user` "fixes" the lack of type. Although you could argue that only users can write, `writable_by?(group)` might make sense. But you could counterattack saying that you should check a group member, not the group itself. Up to you, you can ignore this speech 😅

File: app/controllers/carto/api/state_presenter.rb (hunk 6 above)
Comment (juanignaciosl) [inline app/controllers/carto/api/state_presenter.rb]: Are we renaming `to_poro` to `to_hash`? 😅

File: app/controllers/carto/api/state_presenter.rb [hunk 17]
```diff
@@ -0,0 +1,19 @@
+module Carto
+  module Api
+    class StatePresenter
+      def initialize(state)
+        @state = state
+      end
+
+      def to_hash
```
Comment (juanignaciosl) [inline app/controllers/carto/api/state_presenter.rb]: Are we renaming `to_poro` to `to_hash`? 😅

File: db/migrate/20160808114841_create_states.rb (hunk 2 above)
Comment (javitonino) [inline db/migrate/20160808114841_create_states.rb]: I disagree. Even if it is handled by PG, it still needs to get all the states from that user somehow. If we have millions of states (more than one per viz) and PG needs to do a table scan (order of minutes) each time a user is deleted, it is going to be bad. The current index is even worse as it would do that for each visualization deletion.

As I understand it, we will have queries:
 - Search by PK (visualization state)
 - Search by viz+user_id pair (retrieve snapshots)
 - Seach by viz (viz deletion)
 - Search by user (user deletion)

Sure, the least common of those cases is the search by user, and I agree that the best two column index is the one you proposed (also because viz id is more specific and should go first because of that). But I think having a full scan on a table of millions of rows is still bad news even if it's only for user deletion.

File: app/controllers/carto/builder/public/embeds_controller.rb [hunk 18]
```diff
@@ -35,6 +37,14 @@ def load_visualization
           render_404 unless @visualization
         end
 
+        def load_visualization_for_vizjson_and_state
+          @visualization_for_vizjson_and_state = if @visualization.mapcapped?
```
Comment (juanignaciosl) [inline app/controllers/carto/builder/public/embeds_controller.rb]: Maybe this `if` should be moved to `Carto::Visualization#visualization_for_vizjson`, as the check and the two expression values belong to visualization. But you're The Mapcaps Master™, so up to you :-)

File: app/services/carto/visualizations_export_persistence_service.rb [hunk 19]
```diff
@@ -59,6 +59,11 @@ def save_import(user, visualization, renamed_tables: {})
...
+      state.visualization = visualization
+      state.user = visualization.user
```
Comment (juanignaciosl) [inline app/services/carto/visualizations_export_persistence_service.rb]: Uhm. Not sure about state export/import. Isn't state dependent on ids or usernames?

File: app/services/carto/visualizations_export_service_2.rb [hunk 20]
```diff
@@ -67,6 +68,12 @@ def build_visualization_from_hash(exported_visualization)
         visualization.user = Carto::User.new(username: exported_user[:username])
       end
 
+      # This is optional as it was added in version 2.0.3
+      exported_state = exported_visualization[:state]
+      if exported_state
+        visualization.state = build_state_from_hash(exported_state)
```
Comment (juanignaciosl) [inline app/services/carto/visualizations_export_service_2.rb]: I was thinking about initializing it to empty, but I think that the `alias` trick would do it, isn't it?

File: app/services/carto/visualizations_export_service_2.rb [hunk 21]
```diff
@@ -169,6 +176,12 @@ def build_widget_from_hash(exported_widget, order:, layer:)
         source_id: exported_widget[:source_id]
       )
     end
+
+    def build_state_from_hash(exported_state)
+      return nil unless exported_state
```
Comment (juanignaciosl) [inline app/services/carto/visualizations_export_service_2.rb]: `nil` or `{}`?

File: db/migrate/20160808114842_add_state_id_to_visualizations.rb [hunk 22]
```diff
@@ -0,0 +1,14 @@
...
+    end
+  end
+
+  down do
+    alter_table :visualizations do
+      drop_constraint :visualizations_state_id_fkey
+      drop_column :state_id
```
Comment (juanignaciosl) [inline db/migrate/20160808114842_add_state_id_to_visualizations.rb]: I'm 99% sure that you can drop the column and the FK will be dropped automatically.

File: spec/requests/carto/api/states_controller_spec.rb [hunk 23]
```diff
@@ -0,0 +1,78 @@
...
+
+  let(:state) do
+    {
+      "map": {
+        "center": [40.59896507038094, -73.69937896728516],
+        "zoom": 12
+      },
+      "widgets":
+      {
+        "314accaf-61e7-4a37-803d-c1dd6ffaafed": {
```
Comment (juanignaciosl) [inline spec/requests/carto/api/states_controller_spec.rb]: This makes my doubts certain: exporting/importing won't work well with states without deeper changes.

File: db/migrate/20160808114841_create_states.rb (hunk 2 above)
Comment (juanignaciosl) [inline db/migrate/20160808114841_create_states.rb]: Ok, `visualization_id, user_id` then 👍

Comment (javitonino) [inline db/migrate/20160808114841_create_states.rb]: I'd still go for two separate indexes in order to have index support for the user deletion.

File: db/migrate/20160808114841_create_states.rb (hunk 15 above)
Comment (gfiorav) [inline db/migrate/20160808114841_create_states.rb]: The only reason for state not to be a column in visualisation, is that it will be reused in snapshots. Nothing to do with mapcaps directly. It will be supported, but as any other vis related thing that is frozen in a mapcap (that is, mostly everything exported in a .carto).

File: db/migrate/20160808114841_create_states.rb (hunk 2 above)
Comment (juanignaciosl) [inline db/migrate/20160808114841_create_states.rb]: The `(visualization_id, user_id)` index is not _optimal_ but will be used in many operations filtering by `user_id`. [From the PG manual](https://www.postgresql.org/docs/9.5/static/indexes-multicolumn.html):

> A multicolumn B-tree index can be used with query conditions that involve any subset of the index's columns, but the index is most efficient when there are constraints on the leading (leftmost) columns.

IMHO this is enough for not critical operations, but up to you.

Comment (javitonino) [inline db/migrate/20160808114841_create_states.rb]: Oh, I didn't know that it could use indices like that. It does a full index scan, but that should be good enough for user deletion. :+1: to switching the column order. Good discussion :)

File: app/controllers/carto/api/state_presenter.rb (hunk 17 above)
Comment (gfiorav) [inline app/controllers/carto/api/state_presenter.rb]: haha, oh well I don't really like the "Plain Old Ruby Object" bs, and it really is returning a Hash representation of a State so... Any better naming?

File: app/services/carto/visualizations_export_persistence_service.rb (hunk 19 above)
Comment (gfiorav) [inline app/services/carto/visualizations_export_persistence_service.rb]: Yeah, totally overlooked that :/

Let me think of a solution and I'll come back to this.

File: app/services/carto/visualizations_export_service_2.rb (hunk 20 above)
Comment (gfiorav) [inline app/services/carto/visualizations_export_service_2.rb]: Yeah, that's what I intended. I should make it more explicit though.

File: app/services/carto/visualizations_export_service_2.rb (hunk 21 above)
Comment (gfiorav) [inline app/services/carto/visualizations_export_service_2.rb]: A state is an object, so `nil`. A state's JSON is `{}` if empty.

File: spec/requests/carto/api/states_controller_spec.rb (hunk 23 above)
Comment (gfiorav) [inline spec/requests/carto/api/states_controller_spec.rb]: Yup, thinking this one out.

File: app/controllers/carto/api/state_presenter.rb (hunk 17 above)
Comment (juanignaciosl) [inline app/controllers/carto/api/state_presenter.rb]: No, you're right indeed. This is not a battle that I want to fight, whatever you do is fine to me as I don't have to rename methods myself xD

File: app/services/carto/visualizations_export_service_2.rb (hunk 20 above)
Comment (juanignaciosl) [inline app/services/carto/visualizations_export_service_2.rb]: Well... If you come up with something better, great. Otherwise, the trick is useful and clear as long that it's commented.

File: app/services/carto/visualizations_export_service_2.rb (hunk 21 above)
Comment (juanignaciosl) [inline app/services/carto/visualizations_export_service_2.rb]: 👍

**Comment by gfiorav** [comment]
@xavijam please review the front part!

@javitonino could you have a final look? some profound stuff has changed regarding how state is set in visualisation.

File: lib/assets/javascripts/cartodb3/data/map-definition-model.js [hunk 24]
```diff
@@ -35,12 +34,6 @@ module.exports = Backbone.Model.extend({
     this._mapModel = opts.vis.map;
     this._userModel = opts.userModel;
     this._layerDefinitionsCollection = opts.layerDefinitionsCollection;
-
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb3/data/map-definition-model.js]: Perfect

File: lib/assets/javascripts/cartodb3/data/map-definition-model.js [hunk 25]
```diff
@@ -35,12 +34,6 @@ module.exports = Backbone.Model.extend({
...
-    this._initBinds();
-  },
-
-  _initBinds: function () {
-    this._mapModel.on('moveend', _.debounce(this._onMoveEnd, DEBOUNCE_TIME), this);
   },
 
   _onMoveEnd: function () {
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb3/data/map-definition-model.js]: Remove this method too.

File: app/models/carto/state.rb [hunk 26]
```diff
@@ -0,0 +1,39 @@
...
+
+  serialize :json, ::Carto::CartoJsonSymbolizerSerializer
+
+  validates :json, carto_json_symbolizer: true
+  validates :visualization, :user, presence: true
+
+  after_initialize :ensure_json
+
+  def repopulate_widget_ids(widgets)
```
Comment (javitonino) [inline app/models/carto/state.rb]: Dead code? :fire:

File: app/models/carto/visualization.rb [hunk 27]
```diff
@@ -413,8 +416,27 @@ def populate_ids(ids_json)
...
+
+  def state
+    super ? super : build_state
+  end
+
   private
 
+  def build_state
+    new_state = Carto::State.create!(user: user, visualization: self)
+    update_attribute(:state_id, new_state.id) if persisted?
```
Comment (javitonino) [inline app/models/carto/visualization.rb]: Doesn't `State.create` fail if the visualization isn't persisted?

File: app/models/carto/visualization.rb [hunk 28]
```diff
@@ -413,8 +416,27 @@ def populate_ids(ids_json)
...
+
+    new_state
+  end
+
+  def save_state_if_needed
+    state.save if persisted? && !state.persisted?
```
Comment (javitonino) [inline app/models/carto/visualization.rb]: This doesn't execute on state updates, only creation. Intended?

File: app/models/carto/visualization.rb [hunk 29]
```diff
@@ -58,12 +58,15 @@ class Carto::Visualization < ActiveRecord::Base
   has_many :analyses, class_name: Carto::Analysis
   has_many :mapcaps, class_name: Carto::Mapcap, dependent: :destroy, order: 'created_at DESC'
 
+  belongs_to :state, class_name: Carto::State
+  after_commit :save_state_if_needed
```
Comment (javitonino) [inline app/models/carto/visualization.rb]: I guess it is needed, but we should try to avoid `after_commit` if possible.

File: app/models/carto/visualization.rb [hunk 30]
```diff
@@ -58,12 +58,15 @@ class Carto::Visualization < ActiveRecord::Base
...
+
   def self.columns
     super.reject { |c| c.name == 'url_options' }
   end
 
   def ==(other_visualization)
-    self.id == other_visualization.id
+    id == other_visualization.id
```
Comment (javitonino) [inline app/models/carto/visualization.rb]: I know this is not your code, but isn't this the default behaviour of AR? Are you :psycho: enough to get rid of this altogether?

Comment (gfiorav) [inline app/models/carto/visualization.rb]: I'm unaware of AR behaviour but:

two non persisted objects are the same for this code 💯

File: app/services/carto/visualizations_export_persistence_service.rb [hunk 31]
```diff
@@ -59,6 +59,14 @@ def save_import(user, visualization, renamed_tables: {})
         map.data_layers.each(&:register_table_dependencies)
       end
 
+      state = visualization.state
+      if state
+        visualization.state = Carto::State.new(visualization: visualization,
+                                               user: user,
+                                               json: state.json)
+      end
+      visualization.save
```
Comment (javitonino) [inline app/services/carto/visualizations_export_persistence_service.rb]: There is a save for the visualization already here, one above (via Carto::Visualization) and one below (Member). Can we reuse the member save to update the state_id?

File: db/migrate/20160808114841_create_states.rb (hunk 2 above)
Comment (javitonino) [inline db/migrate/20160808114841_create_states.rb]: Didn't we agree to reverse the index order to [vis, user]?

**Comment by gfiorav** [comment]
OK, made some changes, please CR @javitonino 

this is going in on Monday anyway.

File: app/models/carto/visualization.rb [hunk 32]
```diff
@@ -413,8 +416,35 @@ def populate_ids(ids_json)
...
+  def state
+    super ? super : build_state
+  end
+
   private
 
+  def build_state
+    new_state = Carto::State.new(user: user, visualization: self)
+
+    new_state
```
Comment (javitonino) [inline app/models/carto/visualization.rb]: Unneeded assignment. Also, don't you need to set `state_id` here? It seems to me like this would create an state but not assign it to the visualization.

File: app/models/carto/visualization.rb [hunk 33]
```diff
@@ -413,8 +416,35 @@ def populate_ids(ids_json)
...
+  end
+
+  def save_state_if_needed
+    current_state = state
+
+    if current_state.changed?
+      current_state.visualization = self unless current_state.visualization
+      current_state.user = user unless current_state.user
+
+      if current_state.save!
```
Comment (javitonino) [inline app/models/carto/visualization.rb]: I think you want `save` without exclamation here as you are checking the return code. The exclamation version raises on error.

File: app/services/carto/visualizations_export_persistence_service.rb [hunk 34]
```diff
@@ -59,6 +59,8 @@ def save_import(user, visualization, renamed_tables: {})
         map.data_layers.each(&:register_table_dependencies)
       end
 
+      visualization.save
```
Comment (javitonino) [inline app/services/carto/visualizations_export_persistence_service.rb]: Where did the state assignment go? I guess it is somehow handled by the new, awesomer autosave. In this case, is this save necessary having another save a few lines up (just before layers)?

File: spec/requests/carto/api/states_controller_spec.rb [hunk 35]
```diff
@@ -0,0 +1,78 @@
...
+    @intruder = FactoryGirl.create(:carto_user)
+
+    @map, @table, @table_visualization, @visualization = create_full_visualization(@user)
+    @state = Carto::State.create(visualization_id: @visualization.id, user_id: @user.id)
+  end
+
+  before(:each) { bypass_named_maps }
+
+  after(:all) do
+    Carto::FeatureFlag.destroy_all
```
Comment (javitonino) [inline spec/requests/carto/api/states_controller_spec.rb]: Can you avoid doing this? Just deleting the created FF should work, right? This may cause problems in parallel execution.

File: app/models/carto/visualization.rb (hunk 32 above)
Comment (gfiorav) [inline app/models/carto/visualization.rb]: I'm building, not creating! no ids yet!

File: app/models/carto/visualization.rb (hunk 33 above)
Comment (gfiorav) [inline app/models/carto/visualization.rb]: Yeah, leftover

File: app/services/carto/visualizations_export_persistence_service.rb (hunk 34 above)
Comment (gfiorav) [inline app/services/carto/visualizations_export_persistence_service.rb]: The change is: now when you "autosave" the state, it grabs the visualizaiton's user and vis if not set automatically.

In this case, when building visualization from hash earlier, a `visualization.state = Carto::State.new(json: ...)` was already issued, so the rest of the work is done when saving the vis. Your probably right about the earlier save, I'm trying that out.

File: spec/requests/carto/api/states_controller_spec.rb (hunk 35 above)
Comment (gfiorav) [inline spec/requests/carto/api/states_controller_spec.rb]: Sure!

**Comment by gfiorav** [comment]
Final cr @javitonino please!

**Comment by javitonino** [comment]
Looks like a :package:, feels like a :package:, smells like a :package: -> :shipit:
